File: answers/other/answers/lintcode/Binary Tree Maximum Path Sum.cpp
//@level Medium 
//@type Divide and Conquer Dynamic Programming Recursion
//@result Accepted Total Runtime: 370 ms 100% test cases passed.

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: An integer
     */
    int maxPathSum(TreeNode *root) {
        // write your code here
		if (NULL == root) {
			return INT_MIN / 10;
		}
		int left = getPath(root->left);
		int right = getPath(root->right);
		int my_max = max(root->val, max(root->val + left, max(root->val + right, root->val + left + right)));
		return max(my_max, max(maxPathSum(root->left), maxPathSum(root->right)));
    }
	int getPath(TreeNode *root) {
		if (path_map.end() == path_map.find(root)) {
			if (NULL == root) {
				path_map[NULL] = INT_MIN / 10;
			}
			else {
				path_map[root] = max(root->val, max(root->val + getPath(root->left), root->val + getPath(root->right)));
			}
		}
		return path_map[root];
	}
	map<TreeNode *, int> path_map;
};

File: answers/leetcode/Binary Tree Maximum Path Sum/Binary Tree Maximum Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    private int maxSum;
    
    public int maxPathSum(TreeNode root) {
        this.maxSum = Integer.MIN_VALUE;
        this.maxChildSum(root);
        return this.maxSum;
    }
    
    private int maxChildSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = this.maxChildSum(root.left);
        int right = this.maxChildSum(root.right);
        this.maxSum = Math.max(this.maxSum, root.val);
        this.maxSum = Math.max(this.maxSum, root.val + left);
        this.maxSum = Math.max(this.maxSum, root.val + right);
        this.maxSum = Math.max(this.maxSum, root.val + left + right);
        int childSum = root.val;
        childSum = Math.max(childSum, root.val + left);
        childSum = Math.max(childSum, root.val + right);
        return childSum;
    }
}